Update surface when cycleway:surface and footway:surface match

mnalis opened the pull request:
(or remove `surface=*` if they _don't_ match)

[Compiles OK](https://github.com/mnalis/StreetComplete/actions/runs/3315165654) and seems to work as expected, e.g.:

- when quests answer with `cycleway:surface` same as `footway:surface` (or other way around), it also sets `surface=*` to the same value
- if  `cycleway:surface` and  `footway:surface` are different, it removes `surface=*` tag (if existing)

Closes: https://github.com/streetcomplete/StreetComplete/issues/4526

westnordost commented:
Does this work? Better add unit tests!

mnalis commented:
(pinging @matkoniecz as original author of the quest for his opinion too; but everyone is invited and welcomed to give their opinion)

Helium314 commented:
> is that about null handling of foot (i.e. if foot is not defined, foot != no evaluates to true?)

Yes, `foot != no` is true when there is no `foot`.

Helium314 commented:
This will remove `surface=paved` e.g. if cycleway surface is asphalt and footway surface is paving stones. Not sure if this is desired...

Helium314 commented:
I think that adding `paved` is not relevant, but there is not much complexity involved.
You simply need to
```kotlin
        if (tags["cycleway:surface"] != null && tags["footway:surface"] != null) {
            if (tags["footway:surface"] != tags["cycleway:surface"] && tags["footway:surface"] in ANYTHING_FULLY_PAVED && tags["cycleway:surface"] in ANYTHING_FULLY_PAVED) {
                tags["surface"] = "paved"
            }
        }
```

mnalis commented:
Thanks @Helium314 ; @matkoniecz also contacted me with information that most of that code/constants has already been implemented in https://github.com/matkoniecz/Zazolc/blob/overlays/app/src/main/java/de/westnordost/streetcomplete/osm/surface/Surface.kt#L121 which should hopefully be submitted soon, so I think I'll avoid duplicating that code for now.

So the general idea here is to fix the two failing tests (or rather, improve on the quest logic?) in this PR so it can get in, and then after @matkoniecz submits and merges his code, those constants can be reused for those quests too in another PR (unless someone objects to the idea)

mnalis commented:
Ok, tests are ready and [pass now](https://github.com/mnalis/StreetComplete/actions/runs/3357973659), so ready for next review.

Note that extensive automated tests I've added discovered few [inconsistencies with _**previous**_ code](https://github.com/streetcomplete/StreetComplete/pull/4548#issuecomment-1290703523) (mostly in handling access permissions), so I've fixed those too.

westnordost commented:
> Note that extensive automated tests I've added discovered few https://github.com/streetcomplete/StreetComplete/pull/4548#issuecomment-1290703523 (mostly in handling access permissions), so I've fixed those too.

Could you please put this into an own PR?

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/quests/surface/AddCyclewayPartSurface.kt:
```diff
@@ -25,7 +25,11 @@ class AddCyclewayPartSurface : OsmFilterQuestType<SurfaceAnswer>() {
           )
         )
-        and (access !~ private|no or (foot and foot !~ private|no) or (bicycle and bicycle !~ private|no))
+        and (
+          !access and bicycle !~ private|no
+          or access and access !~ private|no and bicycle !~ private|no
+          or access and access  ~ private|no and bicycle and bicycle !~ private|no
+        )
```
Why this?

mnalis replied:
This addresses second bullet point of this comment: https://github.com/streetcomplete/StreetComplete/pull/4548#issuecomment-1290703523

Basically, the access logic didn't work correctly in original code (i.e. pre-this PR).
For example, the `cycleway surface` was still being asked for `bicycle=private` etc. (which, as I understand it, shouldn't happen).

So I've written in way that is:
- actually working correctly (passes the tests that check for all combinations)
- more readable (IMHO), as it checks each general `access=*` state separately on each line
- (taking a little more space)

mnalis replied:
> quests in general are asked for ways that are either accessible by default, on foot or with the vehicle type you ought to use there

I agree (and that is what I've tried to do), but the English sentence above have the similary problem the code had; and it is interaction between those **cannot** be defined with simple **"or"** (e.g. `foot=no` overrides `access=yes`).

> How much of a theoretical case is this, though?

That I don't know (and I worry it is unlikely that ad-hoc overpass queries trying to find out would survive such attempts on word-wide queries)


> I am still struggling what exactly the cited section of code filters.... would it not evaluate to false for a way that has access=no but which you can access on foot?

It would evaluate to `false`, and correctly IMHO, because `foot` does not matter at all in `cycleway surface` quest. (You qouted `AddCyclewayPartSurface.kt` here!). There is very similar (but with key names changed) code in `AddFootwayPartSurface.kt` where your example of `access=no+foot=yes` would evaluate to `true`:

```
          !access and foot !~ private|no
          or access and access !~ private|no and foot !~ private|no
          or access and access  ~ private|no and foot and foot !~ private|no
```

The new logic seems quite readable to me:
- if general `access` is not defined, check if `foot` specifically is disallowed
- if general `access` is allowed, check if `foot` overrides this
- if general `access` is disallowed, check if `foot` overrides this

There might be **_shorter_** ways to write this, but much **_less readable_** (in my opinion). I found original code not only give incorrect results, but was also treacherous/misleading and confusing (partly due to how SC handles `null` values. YMMV)

> Also, for consistency, such change necessitates also changing dozens of other quest filters to be consistent (hence my initial knee-jerk reaction that this should be an own PR, as it has the potential to grow out of control).

Well I don't disagree that it should be solved globally, but would rather not assign such a task to _myself_ at this time. I've added the automated tests because you've asked me to, and then tried to handle situation when they failed (by fixing incorrect quest code). 

---

Anyway, let me know how you would prefer this to be resolved. IMHO, the current solution is the best one (I'm obviously biased), but if you want me to revert to old access-handling code and remove/comment-out the tests finding inconsistencies (or some other solution), let me know and I'll try to oblige.

westnordost replied:
I prefer to not have that in this PR then. If this is done, it should be done consistently across all quests where such a filter or a similar filter exists. But I am not at all convinced about either that it is supposedly easier to read, that it covers a relevant case or even that the logic would be correct.  As written in my last post.

mnalis replied:
OK, done as requested (reverted to old access checking code, and commented out the tests that would be failing)

westnordost commented:
Or never mind, that will be too much work.

westnordost commented:
Please do not comment out tests. We have version control. If you want to keep those tests around in case a PR like this is accepted, please put them in some branch.

mnalis commented:
OK, removed failing tests completely. Ready for review.

westnordost reviewed (approved):
Looks good to me.

I prefer if maps with more than 1-2 values are defined like this

```diff
- assertIsNotApplicable("highway" to "path", "bicycle" to "designated", "segregated" to "yes", "cycleway:surface" to "unpaved", "note:cycleway:surface" to "it's complicated")
+ assertIsNotApplicable(
+    "highway" to "path",
+    "bicycle" to "designated",
+    "segregated" to "yes",
+    "cycleway:surface" to "unpaved",
+    "note:cycleway:surface" to "it's complicated"
+)
```
for readability but I won't require you to change this to merge this PR.

The test names are all easy to read and descriptive about what they test.

Especially the part about applicability I find a bit unnecessarily extensive, but now that you've already written it, might as well keep it.
